### PR TITLE
pybind: add assert_exists()

### PR DESCRIPTION
**Pull request by zjcmszh**
Signed-off-by: Zhang Jiao <zhangjiao@cmss.chinamobile.com>

### CODE REVIEW

**Comment by zjcmszh** [comment]
@batrick   Please review. Thanks.

**Comment by zjcmszh** [comment]
@tchaikov  Please review, thanks.

**Comment by zjcmszh** [comment]
@tchaikov  please check
Test Code：
```
from rados import (Rados, WriteOpCtx, ObjectNotFound)
from nose.tools import eq_ as eq, ok_ as ok, assert_raises
import threading
import rados
import unittest

class TestIoctx(unittest.TestCase):
    def setUp(self):
        self.rados = Rados(conffile='./ceph.conf')
        self.rados.connect()
        self.rados.create_pool('test_pool')
        assert self.rados.pool_exists('test_pool')
        self.ioctx = self.rados.open_ioctx('test_pool')
    def tearDown(self):
        self.ioctx.close()
        self.rados.delete_pool('test_pool')
        self.rados.shutdown()

    def test_assert_exists(self):
        with WriteOpCtx() as write_op:
            write_op.new(0)
            write_op.assert_exists()
            self.ioctx.operate_write_op(write_op, 'abc')
            eq(self.ioctx.read('abc'), b'')

if __name__ == '__main__':
    unittest.main()

```
Test Results:
```
[root@jiao build]# python test_assert_exists.py 
.
----------------------------------------------------------------------
Ran 1 test in 2.315s

OK

```

File: src/pybind/rados/rados.pyx [hunk 1]
```diff
@@ -298,6 +298,7 @@ cdef extern from "rados/librados.h" nogil:
     void rados_write_op_remove(rados_write_op_t write_op)
     void rados_write_op_truncate(rados_write_op_t write_op, uint64_t offset)
     void rados_write_op_zero(rados_write_op_t write_op, uint64_t offset, uint64_t len)
+    void rados_write_op_assert_exists(rados_write_op_t write_op)
```
Comment (tchaikov) [inline src/pybind/rados/rados.pyx]: @zjcmszh could you order the declaration alphabetically?

Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: I'll do it right away.

Comment (tchaikov) [reply src/pybind/rados/rados.pyx]: thanks

Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: thanks yoursuggestion , updated.

Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: thanks your suggestion , updated.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-11-13_13:53:42-rados-wip-kefu-testing-2019-11-13-1728-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-11-15_11:39:33-rados-wip-kefu-testing-2019-11-13-1728-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-11-15_17:09:33-rados-wip-kefu-testing-2019-11-13-1728-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-11-16_10:02:59-rados-wip-kefu-testing-2019-11-16-1438-distro-basic-mira/

---

- the test.sh failures are fixed by #31689
- test_cls_rbd.sh already tracked by https://tracker.ceph.com/issues/42333, and it's fixed in latest master.
- test_diskprediction_local tracked by https://tracker.ceph.com/issues/42764

**Comment by zjcmszh** [comment]
@tchaikov  I've got it. Thanks.

File: src/test/pybind/test_rados.py [hunk 2]
```diff
@@ -444,6 +444,13 @@ def test_snap_read(self):
         self.ioctx.remove_snap("snap1")
         self.ioctx.remove_object("inhead")
 
+    def test_assert_exists(self):
+        with WriteOpCtx() as write_op:
+            write_op.new(0)
+            write_op.assert_exists()
+            self.ioctx.operate_write_op(write_op, 'abc')
```
Comment (tchaikov) [inline src/test/pybind/test_rados.py]: could you also test a negative case where the object does not exist? also, please make sure the created object(s) is removed after the test finishes.

Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: Yes, I'll test it right away.

Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: Updated，please check it，thanks.

Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: if there is any problem, please tell me and I will modify it.

**Comment by zjcmszh** [comment]
@tchaikov   I want to ask you a question, should i solve this branch conflict?

**Comment by tchaikov** [comment]
> @tchaikov I want to ask you a question, should i solve this branch conflict?

yeah, please do so.

**Comment by zjcmszh** [comment]
@tchaikov  updated please check.

**Comment by tchaikov** [comment]
@zjcmszh could you squash these two commits into a single one and fix your "Signed-off-by" line

because 

> `Signed-off-by: Zhang Jiao zhangjiao@cmss.chinamobile.com`

should be

> `Signed-off-by: Zhang Jiao <zhangjiao@cmss.chinamobile.com>`

and this change is good to merge.

**Comment by tchaikov** [comment]
retest this please.

**Comment by zjcmszh** [comment]
I go home at night and fix it

**Comment by zjcmszh** [comment]
@tchaikov  updated, please check, thanks a lot

**Comment by ideepika** [comment]
jenkins test make check arm64

**Comment by tchaikov** [comment]
@zjcmszh please fix the prefix of the title of the commit message.

**Comment by zjcmszh** [comment]
@tchaikov  updated，thanks

**Comment by ideepika** [comment]
> @zjcmszh please fix the prefix of the title of the commit message.

@zjcmszh commit message still needs an update.

File: src/test/pybind/test_rados.py [hunk 3]
```diff
@@ -444,6 +444,17 @@ def test_snap_read(self):
...
+            eq(self.ioctx.read('abc'), b'')
+            write_op.remove()
+            self.ioctx.operate_write_op(write_op, "abc")
+            with assert_raises(ObjectNotFound):
+                self.ioctx.read('abc')
```
Comment (tchaikov) [inline src/test/pybind/test_rados.py]: @zjcmszh could you help understand how this is related to a negative case of `write_op.assert_exists()`? i mean, even if `write_op.assert_exists()` does not fail, `self.ioctx.read('abc')` can trigger the exception as well.

**Comment by zjcmszh** [comment]
@ideepika  updated，please check.

File: src/test/pybind/test_rados.py (hunk 3 above)
Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: @tchaikov  thank you for your suggestion, I'll figure out how to fix it

**Comment by zjcmszh** [comment]
@ideepika   updated, please check, thanks

File: src/test/pybind/test_rados.py (hunk 3 above)
Comment (tchaikov) [reply src/test/pybind/test_rados.py]: @zjcmszh ping?

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-12-26_05:49:41-rados-wip-kefu-testing-2019-12-26-1055-distro-basic-smithi/

failures tracked / fixed by

- https://tracker.ceph.com/issues/38539
- https://tracker.ceph.com/issues/43396
- https://github.com/ceph/ceph/pull/32427
- https://github.com/ceph/ceph/pull/32426
- https://github.com/ceph/ceph/pull/32428
- #32429

**Pull request closed without merging**